expression/builtin_time.go

```diff
@@ -2256,160 +2287,420 @@ func isDuration(str string) bool {
+func (b *builtinAddDurationAndStringSig) evalDuration(row []types.Datum) (types.Duration, bool, error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDuration, isNull, errors.Trace(err)
 	}
-	switch tp := args[0].Kind(); tp {
-	case types.KindMysqlTime:
-		arg0 := args[0].GetMysqlTime()
-		tmpDuration := arg0.Add(arg1)
-		result, err := tmpDuration.ConvertToTime(arg0.Type)
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		d.SetMysqlTime(result)
-	case types.KindMysqlDuration:
-		arg0 := args[0].GetMysqlDuration()
-		result, err := arg0.Add(arg1)
+	s, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDuration, isNull, errors.Trace(err)
+	}
+	arg1, err := types.ParseDuration(s, getFsp(s))
+	if err != nil {
+		return types.ZeroDuration, true, errors.Trace(err)
+	}
+	if !isDuration(s) {
+		return types.ZeroDuration, true, nil
+	}
+	result, err := arg0.Add(arg1)
+	if err != nil {
+		return types.ZeroDuration, true, errors.Trace(err)
+	}
+	return result, false, nil
+}
+
+type builtinAddTimeStringNullSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString eval a builtinAddDurationAndDurationSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddTimeStringNullSig) evalString(row []types.Datum) (string, bool, error) {
+	return "", true, nil
+}
+
+type builtinAddStringAndDurationSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString eval a builtinAddStringAndDurationSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddStringAndDurationSig) evalString(row []types.Datum) (result string, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	var (
+		ss   string
+		arg1 types.Duration
+	)
+	ss, isNull, err = b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	arg1, isNull, err = b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	if isDuration(ss) {
+		result, err = strDurationAddDuration(ss, arg1)
 		if err != nil {
-			return d, errors.Trace(err)
+			return "", true, errors.Trace(err)
 		}
-		d.SetMysqlDuration(result)
-	default:
-		ss, err := args[0].ToString()
+		return result, false, nil
+	}
+	result, err = strDatetimeAddDuration(ss, arg1)
+	if err != nil {
+		return "", true, errors.Trace(err)
+	}
+	return result, false, nil
+}
+
+type builtinAddStringAndStringSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString eval a builtinAddStringAndStringSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime
+func (b *builtinAddStringAndStringSig) evalString(row []types.Datum) (result string, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	var (
+		s, ss string
```

Comment:
ss and s are confusing,
use a better variable name.

Reply:
ok